More JSON error reporting for extend/delete/relative/proportional

ehughsbaird opened the pull request:
#### Summary
Infrastructure "More JSON error reporting for extend/delete/relative/proportional"

#### Purpose of Change
Fixes https://github.com/CleverRaven/Cataclysm-DDA/issues/81025

#### Describe the solution
Report when `delete` has no effect.

Report when a value loaded through `optional` has an `extend/delete/relative/proportional` specified, but it is not used.

Enable using `relative` for some values - specifically those stored in keyed containers, such as `qualities` on items. There were a lot of places doing this already, but it didn't work.

#### Describe alternatives you've considered
I think the delete check is useful in identifying other errors, but it does remove the ability to specify that X should _not_ be the case.

#### Testing
WIP. In particular, mod JSON is not fixed.

#### Additional context
See commits for more details.

Please don't squash the commits.

ehughsbaird committed:
JSON: Report when delete doesn't delete

Unfortunately the only way to add context, which is necessary to
resolve this, is to throw a JsonError and abort loading.

Also have a debugmsg that reports the specific value when it is not
deleted, before the error. Add a bunch of template code to convert the
data into a string in any way it can.

This isn't reported for the bitsets, and it could, but I think it's
lower priority.

ehughsbaird committed:
JSON: Remove 'delete'd entities that do not delete

Reported by automated check, delete statements have no effect. Remove or
fix the underlying issue.

ehughsbaird committed:
optional() reports failure for non-standard loads

Report when 'extend/delete/relative/proportional' fail for loads with
optional. Values loaded in other ways will not report this.

All game data loading should be converted to use mandatory/optional.

ehughsbaird committed:
Fix exodii ammo damage 'extend'

ehughsbaird committed:
Fix astyle monstrosity in flexbuffer_json.h

astyle is confused and adding a bunch of extra indentation. Fix the
indent and turn it off for that section.

ehughsbaird committed:
Enable loading with relative for keyed containers

There are many pieces of game data that used relative on containers, but
it was not supported. Support it.

ehughsbaird committed:
Generic factory readers object handling

Allow using generic factory readers on objects, if the reader opts into
it.

This can be used for the native representation of a map in json.

ehughsbaird committed:
Generic factory map reader

Not used by anything, so hopefully it doesn't just rot away.

+ can use extend/delete
- keys that can be strings but cannot be constructed from a string

RenechCDDA commented:
I believe 1e43b1652a7074f8a7e5526127a99c094b6e9869 would resolve #81025 (assuming that is not user error)

ehughsbaird committed:
Fix data errors in mods

Revealed by extra error messages for extend/delete/relative/proportional

ehughsbaird committed:
Use readers for containers

Some mods use extend/delete on these containers, but because they didn't
use readers it wasn't supported.

ehughsbaird committed:
JSON: report when extend doesn't extend

Insertion into maps fails when the map already has a corresponding key.
Detect and report this error.

ehughsbaird committed:
Fix material definitions

Duplicate entries are ignored, convert to an object with the desired
weights instead.

ehughsbaird committed:
Fix mod faulty data

These weren't working because they were duplicating an existing key in
the data structure. Fix them.

ehughsbaird commented:
It is a little bit of both.
